cost-model/src/cost_model.rs

```diff
@@ -78,35 +65,82 @@ impl CostModel {
             TransactionCost::SimpleVote { transaction }
         } else {
-            let signature_cost = Self::get_signature_cost(transaction, feature_set);
-            let write_lock_cost = Self::get_write_lock_cost(transaction, feature_set);
-
-            let instructions_data_cost = Self::get_instructions_data_cost(transaction);
-            let allocated_accounts_data_size =
-                Self::calculate_allocated_accounts_data_size(transaction);
-
-            let programs_execution_cost = actual_programs_execution_cost;
+            let num_write_locks = Self::num_write_locks(transaction, feature_set);
             let loaded_accounts_data_size_cost = Self::calculate_loaded_accounts_data_size_cost(
                 actual_loaded_accounts_data_size_bytes,
                 feature_set,
             );
+            let instructions_data_cost =
+                Self::get_instructions_data_cost(transaction.program_instructions_iter());
 
-            let usage_cost_details = UsageCostDetails {
+            Self::calculate_non_vote_transaction_cost(
                 transaction,
-                signature_cost,
-                write_lock_cost,
-                data_bytes_cost: instructions_data_cost,
-                programs_execution_cost,
+                transaction.program_instructions_iter(),
+                num_write_locks,
+                actual_programs_execution_cost,
                 loaded_accounts_data_size_cost,
-                allocated_accounts_data_size,
-            };
+                instructions_data_cost,
+                feature_set,
+            )
+        }
+    }
 
-            TransactionCost::Transaction(usage_cost_details)
+    /// Return an estimated total cost for a transaction given its':
+    /// - `meta` - transaction meta
+    /// - `instructions` - transaction instructions
+    /// - `num_write_locks` - number of requested write locks
+    pub fn estimate_cost<'a, Tx: StaticMeta>(
+        transaction: &'a Tx,
+        instructions: impl Iterator<Item = (&'a Pubkey, SVMInstruction<'a>)> + Clone,
+        num_write_locks: u64,
+        feature_set: &FeatureSet,
+    ) -> TransactionCost<'a, Tx> {
+        if transaction.is_simple_vote_transaction() {
+            return TransactionCost::SimpleVote { transaction };
         }
+        let (programs_execution_cost, loaded_accounts_data_size_cost, data_bytes_cost) =
+            Self::get_transaction_cost(transaction, instructions.clone(), feature_set);
+        Self::calculate_non_vote_transaction_cost(
+            transaction,
+            instructions,
+            num_write_locks,
+            programs_execution_cost,
+            loaded_accounts_data_size_cost,
+            data_bytes_cost,
+            feature_set,
+        )
+    }
+
+    fn calculate_non_vote_transaction_cost<'a, Tx: StaticMeta>(
+        transaction: &'a Tx,
+        instructions: impl Iterator<Item = (&'a Pubkey, SVMInstruction<'a>)> + Clone,
+        num_write_locks: u64,
```

Comment:
nit: this parameter stands out as 'count' while the rest of parameters are 'cost'. Maybe change it to `cost' as well?